src/detect-ftp-mode.c

```diff
@@ -0,0 +1,166 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ *
+ * \author Jeff Lucovsky <jlucovsky@oisf.net>
+ *
+ * Implements the ftp.mode sticky buffer
+ *
+ */
+
+#include "suricata-common.h"
+
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+
+#include "rust.h"
+#include "flow.h"
+
+#include "util-debug.h"
+
+#include "app-layer.h"
+#include "app-layer-ftp.h"
+
+#include "detect-ftp-mode.h"
+
+#define KEYWORD_NAME "ftp.mode"
+#define KEYWORD_DOC  "ftp-keywords.html#ftp-mode"
+#define BUFFER_NAME  "ftp.mode"
+#define BUFFER_DESC  "ftp mode"
+
+typedef struct DetectFtpModeData_ {
+    bool active;
+} DetectFtpModeData;
+
+static int g_ftp_mode_buffer_id = 0;
+
+/**
+ * \brief This function is used to check matches from the FTP App Layer Parser
+ *
+ * \param t pointer to thread vars
+ * \param det_ctx pointer to the pattern matcher thread
+ * \param p pointer to the current packet
+ * \param m pointer to the sigmatch
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectFtpModeMatch(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags, void *state,
+        void *txv, const Signature *s, const SigMatchCtx *m)
+{
+    FTPTransaction *tx = (FTPTransaction *)txv;
+    if (tx->command_descriptor.command_code == FTP_COMMAND_UNKNOWN) {
+        return 0;
+    }
+    if (!tx->dyn_port) {
+        return 0;
+    }
+
+    const DetectFtpModeData *ftpmoded = (const DetectFtpModeData *)m;
+    return 1 ? ftpmoded->active == tx->active : 0;
+}
+
+/**
+ * \brief this function will free memory associated with DetectFtpModeData
+ *
+ * \param ptr pointer to DetectFtpModeData
+ */
+static void DetectFtpModeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCFree(ptr);
+}
+
+/**
+ * \brief This function is used to parse ftp.mode options passed via ftp.mode keyword
+ *
+ * \param str Pointer to the user provided ftp.mode options
+ *
+ * \retval  pointer to DetectFtpModeData on success
+ * \retval NULL on failure
+ */
+static DetectFtpModeData *DetectFtpModeParse(const char *optstr)
+{
+    DetectFtpModeData *ftpmoded = SCCalloc(1, sizeof(DetectFtpModeData));
+    if (unlikely(ftpmoded == NULL))
+        return NULL;
+
+    char *token = SCParseToken(optstr, strlen(optstr));
+    if (!token) {
+        SCLogError("unable to extract token value from \"%s\"", optstr);
+        goto error;
+    } else if (!strcmp(token, "active")) {
+        ftpmoded->active = true;
+    } else if (!strcmp(token, "passive")) {
+        ftpmoded->active = false;
+    } else {
+        SCLogError("Invalid command value");
+        goto error;
+    }
+
+    SCFreeToken(token);
```

Comment:
This feels overkill to realloc and refree...

Why not a `DetectFtpModeData *SCFtpModeParse(char *optstr)` like `SCDetectU16Parse` ?

Reply:
Your suggestion would return a pointer to an object; that object will also have to be released.

Reply:
You already release `DetectFtpModeData` but with my proposal you have 1 object instead of 2 :-p